### ISSUE TITLE
Best way for EntityDecoder to handle MediaType extensions

### ISSUE THREAD

**Issue by jedesah** [http4s/http4s#525]
Currently, quasar has a "bug" where we use the wrong decoder to decode "precise" vs "readable" json. The distinction between the two is represented using `extensions` in the `MediaType`. So currently, we always use one or the other (depending on how they were combined) because the implementation of `satisfiedBy` in `MediaRange` does not look at `extensions`.

It just struck me that I could define a single decoder for json which looks at the extension and determines in which fashion to decode the result. I think this seems pretty clean and I am left wondering why I did not think of this when I first encountered this problem.

I just wanted to post this issue in order to get feedback on whether this seems like it should be the recommended way of solving this problem as opposed to making the `or` combinator on `EntityDecoder` more "aware" of media-type extensions.

If defining a single `EntityDecoder` seems like the desired way of solving this problem, I think it might make sense to consider altering the signature of `decodeBy` to `(mainType: String, qValue: QValue)` in order for the types to convey that extensions do not play a role in "choosing" decoders.

Thoughts?

**Comment by bryce-anderson** [comment]
That is an interesting question. Thinking about it now my personal preference would be to pass a predicate in the construction of the decoder (for the most generic builder at least) that accepts the `Content-Type` header and returns true/false as to whether it knows how to decode the message. I believe that would solve you issue and make the interface much more generic. This is probably an uncommon case but the current case could be build upon it pretty easily.

There may be problems with this idea. I'll think more about it later as time permits.

**Comment by jedesah** [comment]
I like that idea.

**Comment by rossabaker** [comment]
I like it.  What is it going to do to `.consumes`, and what impact is it going to have on Rho?

**Comment by bryce-anderson** [comment]
There it is! rho will be in big trouble if the `EntityDecoder` no longer says what `MediaRange`s it covers.

I don't think this is a show stopper though: we can still require a list of `MediaRange`s while taking a predicate function. The concern becomes that the two need not be related but I'm OK with that. We can preserve the constructor we have now as well for those who don't need such fine grain control since the predicate just becomes 
```scala
contentType => consumes.exists(_.satisfiedBy(contentType.mediaType))
```
I think the predicate will have to become a new member of the trait which may have unforeseen consequences. 

Any thoughts?

**Comment by jedesah** [comment]
Alternatively, we could change the behavior of `MediaType.satisfiedBy(mediaType: MediaRange)` to the following (as opposed to always returning `true`):

```
val `application/json` = MediaType("application", "json")
`application/json`.withExtensions(Map("mode" -> "precise")).satisfiedBy(`application/json`) must_== false
`application/json`.satisfiedBy(`application/json`.withExtensions(Map("mode" -> "precise"))) must_== true
```

That seems like it would solve the problem we are having in Quasar and require minimal changes to the API. But maybe this is surprising behavior?

I have reservations on having both a `MediaType` and a predicate that need to be coherent without any way to enforce it by means of the type system. I think if that approach is to be taken (which may very well be preferable to the proposal I made above), it would be best to accept something similar to `MediaType/MediaRange` but without `extensions` and then accept a predicate over the extensions. This way it's easy enough to deduce the behavior from the types and there is no danger of incoherence.

**Comment by rossabaker** [comment]
I strongly agree on avoiding incoherence.  I am tired and need more time to think which of @jedesah's proposals I prefer, but I think he's on the right track unless someone thinks of a brilliant solution to liberate rho from `.consumes`.

**Comment by bryce-anderson** [comment]
The most recent example of desired behaviour from @jedesah basically looks like it checks if the decoder extensions are a superset of the content extensions. Is that generally a valid algorithm, eg are there ever cases where the extensions override default behaviour and would fail to decode 'non-extended' input? 

If thats all it takes it would be pretty simple to add: just one more check in `satisfiedBy`.

**Comment by jedesah** [comment]
@bryce-anderson I think the conservative answer would be to say no. I attempted to find some specification of extensions for media types, but failed to find anything relevant.

**Comment by jedesah** [comment]
This appears to be the relevant specification (rfc 2045):

```
Parameters are modifiers of the media subtype, and as such do not
fundamentally affect the nature of the content.  The set of
meaningful parameters depends on the media type and subtype.  Most
parameters are associated with a single specific subtype.  However, a
given top-level media type may define parameters which are applicable
to any subtype of that type.  Parameters may be required by their
defining content type or subtype or they may be optional. MIME
implementations must ignore any parameters whose names they do not
recognize.
```

First observation is that it would appear what we are calling "extensions" are actually called "parameters" in the spec.

In light of this, I think it may be reasonable for `isSatisfiedBy` to behave as I described above.

Alternatively we could:

- create a new composition operator for `EntityDecoder` probably called `bestOf(...)` which inspects all `consumes` and chooses the best decoder. This would be in opposition to `OrDec` which makes a binary decision for each decoder and goes down a list until it finds one that is "suitable"
- Say that extensions (parameters) are the responsibility of the `EntityDecoder` and it's up to it to change it's behavior based on them. If so, I suggest we change the signature of `decodeBy` to make it more clear what the `EntityDecoder` looks at in order to decide whether to decode or not a `Message`. This should be easier once #532 is merged but would still be a little awkward with the current way `MediaRange` is defined and I can't think of a better way than to define a `MediaRangeWithoutParams`

Thoughts?
